rfcs/2020-04-04-2221-kubernetes-integration.md

```diff
@@ -0,0 +1,253 @@
+    kubectl create -f vector-configmap.yaml
+    kubectl create -f vector-daemonset.yaml
+    ```
+
+    * *See [outstanding questions 3, 4, 5, 6, and 7](#outstanding-questions).*
+
+    That's it!
+
+#### Helm Interface
+
+TODO: fill in
+
+## Prior Art
+
+1. [Filebeat k8s integration]
+1. [Fluentbit k8s integration]
+2. [Fluentd k8s integration]
+3. [LogDNA k8s integration]
+4. [Honeycomb integration]
+3. [Bonzai logging operator] - This is approach is likely outside of the scope
+   of Vector's initial Kubernetes integration because it focuses more on
+   deployment strategies and topologies. There are likely some very useful
+   and interesting tactics in their approach though.
+4. [Influx Helm charts]
+
+## Sales Pitch
+
+See [motivation](#motivation).
+
+## Drawbacks
+
+1. Increases the surface area that our team must manage.
+
+## Alternatives
+
+1. Not do this integration and rely solely on external community driven
+   integrations.
+
+## Outstanding Questions
+
+1. What is the minimal Kubernetes version that we want to support. See
+   [this comment][kubernetes_version_comment].
+1. What is the best to avoid Vector from ingesting it's own logs? I'm assuming
+   that my [`kubectl` tutoria](#kubectl-interface) handles this with namespaces?
+   We'd just need to configure Vector to excluse this namespace?
+1. I've seen two different installation strategies. For example, Fluentd offers
+   a [single daemonset configuration file][fluentd_daemonset] while Fluentbit
+   offers [four separate configuration files][fluentbit_installation]
+   (`service-account.yaml`, `role.yaml`, `role-binding.yaml`, `configmap.yaml`).
+   Which approach is better? Why are they different?
+1. Should we prefer `kubectl create ...` or `kubectl apply ...`? The examples
+   in the  [prior art](#prior-art) section use both.
+1. From what I understand, Vector requires the Kubernetes `watch` verb in order
+   to receive updates to k8s cluster changes. This is required for the
+   `kubernetes_pod_metadata` transform. Yet, Fluentbit [requires the `get`,
+   `list`, and `watch` verbs][fluentbit_role]. Why don't we require the same?
+1. What is `updateStrategy` ... `RollingUpdate`? This is not included in
+   [our daemonset][vector_daemonset] or in [any of Fluentbit's config
+   files][fluentbit_installation]. But it is included in both [Fluentd's
+   daemonset][fluentd_daemonset] and [LogDNA's daemonset][logdna_daemonset].
+1. I've also noticed `resources` declarations in some of these config files.
+   For example [LogDNA's daemonset][logdna_daemonset]. I assume this is limiting
+   resources. Do we want to consider this?
+1. What the hell is going on with [Honeycomb's integration
+   strategy][Hoenycomb integration]? :) It seems like the whole "Heapster"
+   pipeline is specifically for system events, but Heapster is deprecated?
+   This leads me to my next question...
+1. How are we collecting Kubernetes system events? Is that outside of the
```

Comment:
Kubernetes system events can be collected using Journald from the worker node.
E.g. these services from systemd-journal can be used
```
docker.service
kubelet.service
node-problem-detector.service
```

Reply:
I see two ways we can collect system events, depending on how we define "system events" and what we want to collect.

Collecting host OS logs from the node is one way of thinking about collecting system events. I've addressed this at https://github.com/timberio/vector/pull/2222#discussion_r404654383, so I won't repeat myself here.

One other way of doing this is capturing the events originating from the state changes at the k8s API. This might be very neat for people who want to correlate the k8s state changes with the log data, for instance, to analyze the characteristics of the rolling rollouts, or to detect when the cluster is stuck reconciling something and generates a lot of internal events, flooding its internal gossip system.

Reply:
Added this to the RFC